lib/helper/check_exposures.dart

```diff
@@ -129,57 +101,114 @@ Future<bool> checkExposures() async {
+    }
   }));
 
   user.lastCheck = DateTime.now();
   await user.save();
 
-  if (exposed != null) {
-    print('Found new exposure!');
-    showExposureNotification(exposed);
+  if (exposedBeacons.isNotEmpty) {
+    print('Found new beacon exposures!');
+    var locations = await matchBeaconsAndLocations(exposedBeacons.values);
+    exposedLocations.addAll(locations);
+  }
+
+  // Save all exposed beacons and locations
+  var exposures = [
+    ...exposedBeacons.values.map((b) => Exposure(b)),
+    ...exposedLocations.values.map((l) => Exposure(l)),
+  ];
+
+  await Future.wait(
+    exposures.map((e) {
+      e.exposure = true;
+      return e.save();
+    }),
+  );
+
+  exposures.sort((a, b) => a.end.compareTo(b.end));
+  var exposure = exposures.isNotEmpty ? exposures.last : null;
+
+  if (exposure != null) {
+    showExposureNotification(exposure);
   }
 
   print('Done checking exposures!');
-  return exposed != null;
+  return exposure;
 }
 
-void showExposureNotification(LocationModel location) async {
-  var timestamp = location.timestamp.toLocal();
+/// Takes a list of beacons and attempts to match them with a recored location within
+/// the provided duration window. This method sets the `location` property on each
+/// `BeaconModel` when a match is found. Additionally it returns the set of locations
+/// that matched against any of the provided beacons.
+///
+/// The default `window` Duration is 10 minutes.
+Future<Map<int, LocationModel>> matchBeaconsAndLocations(
+    Iterable<BeaconModel> beacons,
+    {Duration window}) async {
+  window ??= Duration(minutes: 10);
+  var start = beacons.first.start.subtract(window).toIso8601String();
+  var end = beacons.last.end.add(window).toIso8601String();
+
+  List<LocationModel> locations = await LocationModel.findAll(
+      orderBy: 'id ASC',
+      where:
+          'DATETIME(timestamp) >= DATETIME(?) AND DATETIME(timestamp) <= DATETIME(?)',
+      whereArgs: [start, end]);
+  Map<int, LocationModel> locationMatches = {};
+
+  /// For each Beacon do the following:
+  /// - Find all the locations that fall within the duration window if the Beacon time range.
+  /// - Associate the location closest to the midpoint of the Beacon time range to the Beacon.
+  /// - Mark any matching locations as exposures.
+  beacons.forEach((b) {
+    var start = b.start.subtract(window);
+    var end = b.end.add(window);
+    var midpoint = b.start
+        .add(Duration(milliseconds: end.difference(start).inMilliseconds ~/ 2));
+
+    var matches = locations
+        .where((l) => l.timestamp.isAfter(start) && l.timestamp.isBefore(end))
+        .toList();
+
+    matches.sort((a, b) {
+      var aDiff = midpoint.difference(a.timestamp);
+      var bDiff = midpoint.difference(b.timestamp);
+
+      return aDiff.compareTo(bDiff);
+    });
+
+    if (matches.isNotEmpty) {
+      b.location = matches.first;
+      matches.forEach((l) => locationMatches[l.id] = l);
+    }
+  });
+
+  return locationMatches;
```

Comment:
This could possibly be a `Set`?

Reply:
I'd rather keep it as a map since since it matches the type of `exposedLocations`